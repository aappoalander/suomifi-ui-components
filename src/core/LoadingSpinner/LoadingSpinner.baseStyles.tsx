import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { font, element } from '../theme/reset';

export const baseStyles = (theme: SuomifiTheme) => css`
  ${element(theme)}
  ${font(theme)('bodyTextSmall')}
  padding: 20px;
  &.fi-loadingSpinner {
    display: block;
    text-align: center;

    & svg {
      animation: rotation 1.5s infinite linear;
      display: inline-block;
      width: 40px;
      height: 40px;
    }
    & .fi-loadingSpinner-label {
      font-size: 18px;
      font-weight: 600;
      text-align: center;
      margin-top: 6px;
    }
    &.fi-loadingSpinner-labelAlign-right {
      display: flex;
      align-items: center;

      & .fi-loadingSpinner-label {
        margin-left: 10px;
        margin-top: 0;
        line-height: 1.3em;
      }
    }
  }
  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(359deg);
    }
  }
  @media (prefers-reduced-motion) {
    &.fi-loadingSpinner svg {
      animation: rotation 10s infinite linear;
    }
  }
`;
