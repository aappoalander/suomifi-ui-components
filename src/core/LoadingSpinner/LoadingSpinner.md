```js
import { LoadingSpinner } from 'suomifi-ui-components';
<>
  <div aria-live="polite">
    <LoadingSpinner labelText="Loading data" />
  </div>
  {/*
  <LoadingSpinner labelText="Loading data" labelAlign="right" />
  <LoadingSpinner labelText="Loading data" labelMode="hidden" />
  <LoadingSpinner labelText="Loading data" size="small" />
  <LoadingSpinner
    labelText="Loading data"
    labelAlign="right"
    size="small"
  />
  <LoadingSpinner labelText="Loading successful" status="success" />
  <LoadingSpinner
    labelText="Loading successful"
    labelAlign="right"
    status="success"
  />
  <LoadingSpinner
    labelText="Loading successful"
    labelMode="hidden"
    status="success"
  />
  <LoadingSpinner
    labelText="Loading successful"
    size="small"
    status="success"
  />
  <LoadingSpinner
    labelText="Loading successful"
    labelAlign="right"
    size="small"
    status="success"
  />
  <LoadingSpinner labelText="Loading failed" status="fail" />
  <LoadingSpinner
    labelText="Loading failed"
    labelAlign="right"
    status="fail"
  />
  <LoadingSpinner
    labelText="Loading failed"
    labelMode="hidden"
    status="fail"
  />
  <LoadingSpinner
    labelText="Loading failed"
    size="small"
    status="fail"
  />
  <LoadingSpinner
    labelText="Loading failed"
    labelAlign="right"
    size="small"
    status="fail"
  />
  */}
</>;
```
