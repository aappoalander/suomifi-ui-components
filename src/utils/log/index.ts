export { getLogger, setLogger, Logger } from './logger';
