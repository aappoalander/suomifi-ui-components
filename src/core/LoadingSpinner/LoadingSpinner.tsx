import React, { Component, forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { baseStyles } from './LoadingSpinner.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../theme';
import { HtmlDiv, HtmlDivWithRef } from '../../reset';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden';
import { Icon } from '../../core/Icon/Icon';

export interface LoadingSpinnerProps {
  /** Custom class name for styling and customizing */
  className?: string;
  /** Unique id */
  id?: string;
  /** Label text for LoadingSpinner */
  labelText: string;
  /** Align label on bottom or on the right side of animation svg
   * @default 'bottom'
   */
  labelAlign?: 'bottom' | 'right';
  /** Is LoadingSpinners's Label visible or not
   * @default 'visible'
   */
  labelMode?: 'visible' | 'hidden';
  /** Is LoadingSpinners's status. What it indicates
   * @default 'loading'
   */
  status?: 'loading' | 'success' | 'fail';
  /** Is LoadingSpinners's size. Its it normal or small
   * @default 'loading'
   */
  size?: 'normal' | 'small';
}
export interface LoadingSpinnerState {
  loaded?: number;
}
interface InnerRef {
  forwardedRef: React.RefObject<HTMLDivElement>;
}
const baseClassName = 'fi-loadingSpinner';
export const loadingSpinnerClassNames = {
  label: `${baseClassName}-label`,
  labelAlign: '',
  size: '',
  status: '',
};
export const loadingSpinnerState = {
  loaded: 0,
};
class BaseLoadingSpinner extends Component<
  LoadingSpinnerProps & InnerRef & LoadingSpinnerState
> {
  constructor(props) {
    super(props);

    this.state = {
      loaded: 0,
      visible: false,
    };
  }

  componentDidMount() {
    setTimeout(() => {
      this.setState({
        loaded: 1,
      });
      setTimeout(() => {
        this.setState({
          loaded: 2,
        });
        setTimeout(() => {
          this.setState({
            loaded: 3,
          });
          setTimeout(() => {
            this.setState({
              loaded: 4,
            });
            setTimeout(() => {
              this.setState({
                loaded: 5,
              });
            }, 2000);
          }, 2000);
        }, 2000);
      }, 2000);
    }, 2000);
  } /*
  timeOut = () => {

  } */

  render() {
    const {
      className,
      id,
      labelText,
      labelAlign = 'bottom',
      labelMode = 'visible',
      size = 'normal',
      status = 'loading',
      ...passProps
    } = this.props;
    loadingSpinnerClassNames.labelAlign = `${baseClassName}-labelAlign-${labelAlign}`;
    loadingSpinnerClassNames.size = `${baseClassName}-size-${size}`;
    loadingSpinnerClassNames.status = `${baseClassName}-status-${status}`;

    return (
      <>
        <button
          onClick={() => {
            this.setState({ visible: true });
          }}
        >
          show spinner
        </button>
        {this.state.visible && (
          <HtmlDivWithRef
            className={classnames(
              baseClassName,
              className,
              loadingSpinnerClassNames.labelAlign,
              loadingSpinnerClassNames.size,
              loadingSpinnerClassNames.status,
            )}
            as="section"
            id={id}
            {...passProps}
          >
            {status === 'loading' && (
              <svg aria-hidden="true" viewBox="0 0 40 40" version="1.1">
                <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
                  <path
                    d="M20,0 L20,3 C10.6111593,3 3,10.6111593 3,20 C3,29.3888407 10.6111593,37 20,37 
              C29.280923,37 36.824796,29.5628044 36.9969921,20.3230397 L37,20 L40,20 C40,31.045695 
              31.045695,40 20,40 C8.954305,40 0,31.045695 0,20 C0,9.06936433 8.76872859,0.186774951 
              19.6555106,0.00290705581 L20,0 Z"
                    id="Path"
                    fill="#00347A"
                    fillRule="nonzero"
                  />
                </g>
              </svg>
            )}
            {status === 'success' && <Icon icon="checkCircleFilled" />}
            {status === 'fail' && <Icon icon="errorFilled" />}

            {labelMode === 'visible' ? (
              <HtmlDiv className={loadingSpinnerClassNames.label}>
                {labelText}
              </HtmlDiv>
            ) : (
              <VisuallyHidden>{labelText}</VisuallyHidden>
            )}
            <p> has been loaded {this.state.loaded} / 5</p>
          </HtmlDivWithRef>
        )}
      </>
    );
  }
}
const StyledLoadingSpinner = styled(
  (props: LoadingSpinnerProps & InnerRef & SuomifiThemeProp) => {
    const { theme, ...passProps } = props;
    return <BaseLoadingSpinner {...passProps} />;
  },
)`
  ${({ theme }) => baseStyles(theme)};
`;
export const LoadingSpinner = forwardRef(
  (props: LoadingSpinnerProps, ref: React.RefObject<HTMLDivElement>) => {
    const { ...passProps } = props;
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <StyledLoadingSpinner
            forwardedRef={ref}
            theme={suomifiTheme}
            {...passProps}
          />
        )}
      </SuomifiThemeConsumer>
    );
  },
);
