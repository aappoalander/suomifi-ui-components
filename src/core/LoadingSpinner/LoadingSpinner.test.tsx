import React from 'react';
import { render } from '@testing-library/react';
import { LoadingSpinner } from './LoadingSpinner';
import { axeTest } from '../../utils/test';

describe('props', () => {
  describe('labelMode should be visually-hidden', () => {
    const { getByText } = render(
      <LoadingSpinner labelText="To be hidden" labelMode="hidden" />,
    );
    const spinner = getByText('To be hidden');
    expect(spinner).toHaveClass('fi-visually-hidden');
  });
  describe('id', () => {
    const LoadingSpinnerWithElementId = (
      <LoadingSpinner id="TestId" labelText="Loading data" />
    );
    const { container } = render(LoadingSpinnerWithElementId);
    expect(container.querySelector('#TestId')).toHaveClass('fi-loadingSpinner');
  });
  it('labelAlign', () => {
    const labelAlignLoadingSpinner = (
      <LoadingSpinner labelAlign="right" labelText="Loading data" />
    );
    const { container } = render(labelAlignLoadingSpinner);
    expect(container.firstChild).toHaveClass(
      'fi-loadingSpinner-labelAlign-right',
    );
  });
  describe('className', () => {
    const customClassLoadingSpinner = (
      <LoadingSpinner labelText="Loading data" className="custom-class" />
    );

    it('contains base classname', () => {
      const { container } = render(customClassLoadingSpinner);
      expect(container.firstChild).toHaveClass('fi-loadingSpinner');
    });

    it('contains custom classname', () => {
      const { container } = render(customClassLoadingSpinner);
      expect(container.firstChild).toHaveClass('custom-class');
    });
  });
  it('labelText', () => {
    const LabelTextLoadingSpinner = (
      <LoadingSpinner labelText="Loading data" className="custom-class" />
    );
    const { getByText } = render(LabelTextLoadingSpinner);
    const label = getByText('Loading data');
    expect(label).toHaveClass('fi-loadingSpinner-label');
  });
});

describe('accessibility', () => {
  const TestLoadingSpinner = <LoadingSpinner labelText="Loading data" />;
  test(
    'should not have basic accessibility issues',
    axeTest(TestLoadingSpinner),
  );
});
