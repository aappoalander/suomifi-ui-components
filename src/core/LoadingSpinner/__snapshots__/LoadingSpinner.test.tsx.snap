// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should match snapshot 1`] = `
.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1.fi-loadingSpinner {
  display: block;
  text-align: center;
}

.c1.fi-loadingSpinner svg {
  display: inline-block;
  width: 40px;
  height: 40px;
}

.c1.fi-loadingSpinner .fi-loadingSpinner-label {
  font-size: 18px;
  font-weight: 600;
  text-align: center;
  margin-top: 6px;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-labelAlign-right {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-labelAlign-right .fi-loadingSpinner-label {
  margin-left: 10px;
  margin-top: 0;
  line-height: 1.3em;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-size-small svg {
  width: 24px;
  height: 24px;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-size-small .fi-loadingSpinner-label {
  font-size: 16px;
  font-weight: normal;
  margin-top: 0;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-status-loading svg {
  -webkit-animation: rotation 1.5s infinite linear;
  animation: rotation 1.5s infinite linear;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-status-success svg path {
  fill: hsl(166,90%,34%);
}

.c1.fi-loadingSpinner.fi-loadingSpinner-status-fail svg path {
  fill: hsl(3,59%,48%);
}

@media (prefers-reduced-motion) {
  .c1.fi-loadingSpinner.fi-loadingSpinner-status-loading svg {
    -webkit-animation: rotation 10s infinite linear;
    animation: rotation 10s infinite linear;
  }
}

<div>
  <section
    class="c0 fi-loadingSpinner c1 fi-loadingSpinner-labelAlign-bottom fi-loadingSpinner-size-normal fi-loadingSpinner-status-loading"
  >
    <svg
      aria-hidden="true"
      version="1.1"
      viewBox="0 0 40 40"
    >
      <g
        fill="none"
        fill-rule="evenodd"
        stroke="none"
        stroke-width="1"
      >
        <path
          d="M20,0 L20,3 C10.6111593,3 3,10.6111593 3,20 C3,29.3888407 10.6111593,37 20,37 
              C29.280923,37 36.824796,29.5628044 36.9969921,20.3230397 L37,20 L40,20 C40,31.045695 
              31.045695,40 20,40 C8.954305,40 0,31.045695 0,20 C0,9.06936433 8.76872859,0.186774951 
              19.6555106,0.00290705581 L20,0 Z"
          fill="#00347A"
          fill-rule="nonzero"
          id="Path"
        />
      </g>
    </svg>
    <div
      class="c2 fi-loadingSpinner-label"
    >
      Loading data
    </div>
  </section>
</div>
`;
